tests/unit_tests/test_gameconnection.py

```diff
@@ -438,56 +442,119 @@ async def test_handle_action_GameEnded_ends_game(
+    time_taken = "09:08:07.654321"
 
-async def test_handle_action_OperationComplete_invalid(ugame: Game, game_connection: GameConnection, database):
-    """
-        Sends an OperationComplete action to handle action and verifies that
-    the `coop_leaderboard` table is updated accordingly.
+    await game_connection.handle_action(
+        "OperationComplete", [False, True, time_taken]
+    )
 
-    Requires that the map from `game.map_file_path` exists in the database.
-    """
+    async with database.acquire() as conn:
+        result = await conn.execute(
+            "SELECT secondary, gameuid from `coop_leaderboard` where gameuid=%s",
+            ugame.id,
+        )
 
+        row = await result.fetchone()
+        assert row is None
+
+
+async def test_handle_action_OperationComplete_invalid(
+    ugame: Game,
+    game_connection: GameConnection,
+    database,
+):
     ugame.map_file_path = "maps/prothyon16.v0005.zip"
     ugame.validity = ValidityState.OTHER_UNRANK
     game_connection.game = ugame
-
-    secondary = 1
     time_taken = "09:08:07.654321"
-    await game_connection.handle_action("OperationComplete", ["1", secondary, time_taken])
+
+    await game_connection.handle_action(
+        "OperationComplete", [True, True, time_taken]
+    )
 
     async with database.acquire() as conn:
         result = await conn.execute(
             "SELECT secondary, gameuid from `coop_leaderboard` where gameuid=%s",
-            ugame.id)
+            ugame.id,
+        )
 
         row = await result.fetchone()
+        assert row is None
 
-    assert row is None
+
+async def test_handle_action_OperationComplete_duplicate(
+    ugame: Game,
+    game_connection: GameConnection,
+    database,
+    caplog,
+):
+    ugame.map_file_path = "maps/prothyon16.v0005.zip"
+    ugame.validity = ValidityState.COOP_NOT_RANKED
+    game_connection.game = ugame
+    time_taken = "09:08:07.654321"
+
+    async with database.acquire() as conn:
+        # OperationComplete expects an existing corresponding game_stats row,
+        # we automatically add such a row for ugame.id == 1 in test-data.sql
+        await conn.execute(
+            game_stats.insert().values(
+                id=ugame.id,
+                startTime=datetime.utcnow(),
+                gameName='Another test game',
+                gameType='0',
+                gameMod=6,
+                host=1,
+                mapId=1,
+                validity=0,
+            )
+        )
+
+    with caplog.at_level(logging.ERROR):
+        await game_connection.handle_action(
+            "OperationComplete", [True, True, time_taken]
+        )
+        caplog.clear()
+        await game_connection.handle_action(
+            "OperationComplete", [True, True, time_taken]
+        )
+
+        assert not any(
+            record.exc_info
+            and isinstance(record.exc_info[1], pymysql.err.IntegrityError)
+            for record in caplog.records
+        )
```

Comment:
Cool, haven't seen this before.

Reply:
If you mean `caplog`, neither had I!

Reply:
Well I knew about caplog, but I didn't know you could use it to check the `exc_info` and get the exception type and stuff.

Reply:
Ah yes, good old `LogRecord`